File: 394-字符串解码.cpp
class Solution {
public:
    string decodeString(string s) {
        string result;
        string subs;
        int subn;
        int j;
        int x;
        
        for(auto i : s){
            if(i != ']')
                result += i;
            else{
                j = result.size();
                subs = "";
                subn = 0;
                x = 0;
                while(isalpha(result[--j]))
                    subs = result[j] + subs;
                while(j > 0 && isdigit(result[--j]))
                    subn += pow(10, x++) * ((int)result[j] - 48);
                result.erase(j > 0 ? j + 1 : 0);
                for(int z = 0; z < subn; z++){
                    result += subs;
                }
            }
        }

        return result;  
    }
};

File: 394-字符串解码.py
class Solution:
    def decodeString(self, s: str) -> str:
        mystack = []
        
        for i in s:
            if i != ']':
                mystack.append(i)
            else:
                subs = str(mystack.pop())
                while mystack[-1].isalpha():
                    subs = str(mystack.pop()) + subs
                mystack.pop()
                subn = int(mystack.pop())
                j = 1
                while mystack and mystack[-1].isnumeric():
                    subn += 10 ** j * int(mystack.pop())
                    j += 1
                mystack.append(subs * subn)
        
        return ''.join(mystack)
